File: app/src/gui/helpers/ImageScannerWorker.cpp
#include "ImageScannerWorker.h"
#include <QDir>
#include <QDirIterator>
#include <QFileInfo>

// Define supported formats (assumed from Python's SUPPORTED_IMG_FORMATS)
const QStringList ImageScannerWorker::SUPPORTED_IMG_FORMATS = {
    "png", "jpg", "jpeg", "bmp", "webp", "gif"
};

ImageScannerWorker::ImageScannerWorker(const QString& directory, QObject* parent)
    : QObject(parent)
{
    // Handle single string input
    if (!directory.isEmpty() && QFileInfo(directory).isDir()) {
        m_directories.append(directory);
    }
}

ImageScannerWorker::ImageScannerWorker(const QStringList& directories, QObject* parent)
    : QObject(parent)
{
    // Filter out empty or non-directory entries
    for (const QString& dir : directories) {
        if (!dir.isEmpty() && QFileInfo(dir).isDir()) {
            m_directories.append(dir);
        }
    }
}

void ImageScannerWorker::runScan()
{
    QStringList imagePaths;
    
    // Create a set of lowercase suffixes, e.g., ".png"
    QSet<QString> supportedSuffixes;
    for (const QString& fmt : SUPPORTED_IMG_FORMATS) {
        supportedSuffixes.insert(QString(".%1").arg(fmt.toLower()));
    }

    if (m_directories.isEmpty()) {
        emit scanError("No valid directories provided for scanning.");
        return;
    }

    try {
        for (const QString& directory : m_directories) {
            // QDirIterator replaces os.walk
            QDirIterator it(directory, QDir::Files, QDirIterator::Subdirectories);
            while (it.hasNext()) {
                QString path = it.next();
                QString suffix = QFileInfo(path).suffix().toLower();
                
                if (!suffix.startsWith('.')) {
                    suffix.prepend('.');
                }

                if (supportedSuffixes.contains(suffix)) {
                    imagePaths.append(QDir::toNativeSeparators(path));
                }
            }
        }
        
        imagePaths.sort();
        emit scanFinished(imagePaths);
        
    } catch (const std::exception& e) {
        emit scanError(QString("Could not scan directory and subdirectories: %1").arg(e.what()));
    }
}

File: gui/helpers/image_scanner_worker.py
import os
from typing import List, Union
from PySide6.QtCore import QObject, Signal, Slot
from backend.src.utils.definitions import SUPPORTED_IMG_FORMATS


class ImageScannerWorker(QObject):
    """
    Worker to perform file system scanning on a separate thread.
    Now supports initialization with a single directory path or a list of paths
    for unified processing (delegating list handling to internal logic).
    """
    
    scan_finished = Signal(list)
    scan_error = Signal(str)

    def __init__(self, directories: Union[str, List[str]]):
        super().__init__()
        
        # --- MODIFICATION: Handle single string or list input ---
        if isinstance(directories, (str, os.PathLike)):
            self.directories = [str(directories)]
        elif isinstance(directories, list):
            # Filter out empty or None entries
            self.directories = [d for d in directories if d and os.path.isdir(d)]
        else:
            self.directories = []
        # --- END MODIFICATION ---

    @Slot()
    def run_scan(self):
        """
        Iterates through all provided directories and aggregates image paths
        using os.walk() for recursive scanning.
        """
        image_paths = []
        supported_extensions = tuple(f'.{fmt}' for fmt in SUPPORTED_IMG_FORMATS)
        
        if not self.directories:
            self.scan_error.emit("No valid directories provided for scanning.")
            return

        try:
            for directory in self.directories:
                if not os.path.isdir(directory):
                    self.scan_error.emit(f"Path is not a valid directory: {directory}")
                    # Note: We emit an error and return immediately, preventing processing of subsequent paths.
                    # For continued processing, this error handling would need to be inside the loop.
                    return

                for root, _, filenames in os.walk(directory):
                    for entry in filenames:
                        # Check for supported image extensions
                        if entry.lower().endswith(supported_extensions):
                            full_path = os.path.join(root, entry) 
                            image_paths.append(full_path)
            
            # Send the sorted results back to the main thread
            self.scan_finished.emit(sorted(image_paths))
        except Exception as e:
            # Handle potential errors like PermissionError
            self.scan_error.emit(f"Could not scan directory and subdirectories: {e}")
